### PR TITLE
#165273472 Add signup activation link

### PR DESCRIPTION
**Pull request by mwinel**
### **What does this PR do?**
This PR adds the capability for the User to receive a verification link via email upon successful signup.

**Tasks**
- Add signup activation link
- Add activation APIView
- Add email_verified field to the User model
- Add URL names for signup and login URLs

### **How to test it?**
Try it out in postman;
1. Add the following to your `.env` file
```
export EMAIL_HOST=smtp.gmail.com
export EMAIL_HOST_USER=youremail@gmail.com
export EMAIL_HOST_PASSWORD=yourpassword
export EMAIL_PORT=587
```
2. Signup
3. Click the activation link
4. The activation link redirects you to a `GET` method with such a `URL` `http://127.0.0.1:8000/api/users/bmVsc29uMTExMjMy/`
5. Response on successful activation
```
{
    "message": "Account successfully activated. Login now."
}
```
6. Response on an unsuccessful activation
```
{
    "error": "Activation link is invalid."
}
```

### **Relevant User Stories**
[#165273472](https://www.pivotaltracker.com/story/show/165273472)

### CODE REVIEW

File: authors/apps/authentication/serializers.py [hunk 1]
```diff
@@ -94,7 +93,7 @@ def validate(self, data):
 class UserSerializer(serializers.ModelSerializer):
     """Handles serialization and deserialization of User objects."""
 
-    # Passwords must be at least 8 characters, but no more than 128 
+    # Passwords must be at least 8 characters, but no more than 128
```
Comment (archibishop) [inline authors/apps/authentication/serializers.py]: @mwinel Remove all unnecessary Changes.

File: authors/apps/authentication/views.py [hunk 2]
```diff
@@ -73,3 +98,24 @@ def update(self, request, *args, **kwargs):
 
         return Response(serializer.data, status=status.HTTP_200_OK)
 
+
+class AccountActivationAPIView(APIView):
```
Comment (3Nakajugo) [inline authors/apps/authentication/views.py:116]: could you please add a doc string

File: .gitignore [hunk 3]
```diff
@@ -40,6 +42,7 @@ pip-delete-this-directory.txt
 htmlcov/
 .tox/
 .coverage
+.coveragerc
```
Comment (sanya-kenneth) [inline .gitignore:45]: Please remove this line because we are not using the .coveragerc file

File: authors/apps/authentication/serializers.py (hunk 1 above)
Comment (mwinel) [reply authors/apps/authentication/serializers.py]: Working on it.

**Review by joelmugaya — APPROVED** [review]

File: authors/apps/authentication/models.py [hunk 4]
```diff
@@ -95,19 +100,18 @@ class User(AbstractBaseUser, PermissionsMixin):
...
 
     @property
     def get_full_name(self):
-      """
-      This method is required by Django for things like handling emails.
-      Typically, this would be the user's first and last name. Since we do
-      not store the user's real name, we return their username instead.
-      """
-      return self.username
+        """
```
Comment (archibishop) [inline authors/apps/authentication/models.py:108]: @mwinel Avoid such unnecessary changes.

Comment (mwinel) [reply authors/apps/authentication/models.py:108]: @archibishop there was an indentation error here that was making the app fail to run.

**Comment by archibishop** [comment]
@mwinel Fix coverage and rebase from develop. Your pr is not building the review on heroku.

File: authors/apps/authentication/tests/test_authentication.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+        self.user_login = {
+            "user": {
+                "email": "mwinel@live.com",
+                "password": "12345678"
+            }
+        }
+
+    def test_create_user(self):
+        """Test create user."""
```
Comment (archibishop) [inline authors/apps/authentication/tests/test_authentication.py:24]: @mwinel This is testing for many things in one test. You should separate them so you follow TDD principles. `A Test should only test for one thing and should be independent.` So you cant have 10 assertions under one test.

File: authors/apps/authentication/tests/test_authentication.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+from rest_framework .test import APITestCase, APIClient
+
+
+class UserManagerTestCase(APITestCase):
+    """Usermanager TestCase."""
+
+    def setUp(self):
+        self.client = APIClient()
+
+        self.user_registration = {
```
Comment (archibishop) [inline authors/apps/authentication/tests/test_authentication.py]: @mwinel Use test fixture. Have a sync with @3Nakajugo so you can proceed with this. I left a similar comment on her pr.

File: authors/apps/authentication/views.py [hunk 7]
```diff
@@ -88,3 +115,28 @@ def update(self, request, *args, **kwargs):
...
+    """
+    Return the registration activation link.
+    """
+    permission_classes = (AllowAny, )
+    serializer_class = UserSerializer
+
+    def get(self, request, uid):
+        try:
+            username = urlsafe_base64_decode(uid).decode('utf-8')
+            user = User.objects.filter(username=username).first()
```
Comment (archibishop) [inline authors/apps/authentication/views.py:127]: @mwinel Confirm this throws the exception(User.DoesNotExist) you expect it to throw.

File: authors/settings.py [hunk 8]
```diff
@@ -157,11 +157,11 @@
     )
 }
 
-#heroku deploy settings
```
Comment (archibishop) [inline authors/settings.py:160]: @mwinel  Return this these lines of code. They are used for heroku.

File: .gitignore (hunk 3 above)
Comment (archibishop) [inline .gitignore:45]: @mwinel this file doesn't exist in the code base. Remove this line.

**Comment by archibishop** [comment]
@mwinel When i `signin`. It doesn't return the activation link. Look into this.

**Comment by mwinel** [comment]
@archibishop The activation link is returned on `signup`. Maybe we can do a check that allows users to only `login` when their accounts are verified, if not they receive a message to prompt them activate their account first.

**Comment by archibishop** [comment]
@mwinel My bad i meant `sign up`

File: .gitignore (hunk 3 above)
Comment (mwinel) [reply .gitignore:45]: We need it incase we are trying to execute a coverage report locally.

**Review by engjames — APPROVED** [review]
@mwinel I have tested your PR and it works as expectedly.

**Review by sanya-kenneth — APPROVED** [review]
